# AbInitioVO and FastFloppyTail
 AbInitio Variable Order and FastFloppyTail Algorithms for Protein Structure Prediction from Sequence

## Algorithm Descritions
### AbInitioVO
The AbInitio Variable Order (AbInitioVO) PyRosetta algorithm is an improvement of the [AbInitio-Relax](https://www.rosettacommons.org/docs/latest/application_documentation/structure_prediction/abinitio-relax)
algorithm in Rosetta which features changes to the fragment selection process and structure scoring. Disordered probability predictions based on protein sequence are used as an additional inputs alongside secondary
structure predictions to influence 9-mer and 3-mer [fragment picking](https://www.rosettacommons.org/docs/latest/application_documentation/utilities/app-fragment-picker). Furthermore, by making improvements to the 
[centroid score terms](https://www.rosettacommons.org/docs/latest/rosetta_basics/scoring/centroid-score-terms), specifically the ""rg"" term, to accomadate disordered proteins this algorithm is able to accurately 
predict the structures of ordered domains, of both fully ordered and partially ordered proteins, as well as identifying disordered regions.

### FastFloppyTail
FastFloppyTail is an improved variant of the [FloppyTail](https://www.rosettacommons.org/docs/latest/application_documentation/structure_prediction/floppy-tail) Rosetta algorithm written for PyRosetta. The algorithm features 
a reduction in sidechain packing moves along with an increase in the fequency with which sampled structures are returned to their minima. Overall this algorithm boasts both a marked increase in accuracy compared to experimental 
data along with a 10-fold reduction in compute time compared to FloppyTail.

## Installation Guide
__Operating System:__ Linux (64-bit)

__Programming Langauge:__ Python
This code was specifically written and tested in Python3.6 (python3.6.8)
	
__Required Python Packages:__
- PyRosetta
	- This was specifically written and tested with PyRosetta 2019.17+release.2cb3f3a py36_0. Additional information can be found on the [PyRosetta](http://www.pyrosetta.org/) website. This specific PyRosetta build can be downloaded [here](http://www.pyrosetta.org/dow) after obtaining a [license](https://els.comotion.uw.edu/express_license_technologies/pyrosetta)
- biopython (1.73)
- numpy (1.14.5)
- scipy (1.1.0)

__Anaconda Environment:__
An anaconda environment containing all necessary packages can be found in the anaconda folder. Build time for this Anaconda environment takes on the order of mintues to hours depending on the number of processors used and is largely dependent on the PyRosetta build time.

__Additional Reccommended Software Packages:__
- Blast (blast-2.2.26)
- [Rosetta](https://www.rosettacommons.org/software)
- [DSSP](https://swift.cmbi.umcn.nl/gv/dssp/) (installed via [conda](https://anaconda.org/salilab/dssp)) 
- [Sparta+](https://spin.niddk.nih.gov/bax/software/SPARTA+/)
- [PALES](https://www3.mpibpc.mpg.de/groups/zweckstetter/_links/software_pales.htm#HuP)

## Running AbInitioVO
AbInitioVO is run using the command-line, where inputs are specified using an arguement parser. An example run which can also be found in the demo section is:
```
run AbInitioVO.py -in asyn.fasta.txt -abinitiovo - abnstruct 5 -diso asyn.diso -t_frag asyn_dcor_frags.200.3mers -n_frag asyn_dcor_frags.200.9mers
```
Each of the parser flags are described below:
```
-in  --Input_FASTA_File  Name of the text file containing the FASTA sequence of the protein of interest. Carot should not be in same line as sequence, UniProt format preferred.
-abnstruct  --Number_of_AbInitio_Structures  Number of structures to sample during Centroid AbInitio portion. Default = 1000
-diso  --Input_DisoPred_File  Name of the file containing the per residue disordered probability prediction generated by RaptorX
-t_frag  --Three_Mer_Frag_Library  Name of the file containing the three-mer fragment library generated by disorder corrected method
-n_frag  --Nine_Mer_Frag_Library  Name of the file containing the nine-mer fragment library generated by disorder corrected method
-rsd_wt_helix  --Residue_Weight_Helix  Reweights the env, pair and cbeta scores for helical residues by specificied factor. Default: 0.5
-rsd_wt_loop  --Residue_Weight_Loop  Reweights the env, pair and cbeta scores for loop residues by specificied factor. Default: 0.5
-rsd_wt_sheet  --Residue_Weight_Sheet  Reweights the env, pair and cbeta scores for sheet residues by specificied factor. Default: 1.0
-rg_weight  --Rg_Weight  Reweights the weight of rg in each step by specificied factor. Default: 0.5
-cycles  --AbInitio_Cycles  Number of sampling cycles within each stage of AbInitio, identical to increase_cycles flag in C++ ClassicAbInitio and AbRelax. Default 10
-abinitiovo  --AbInitioVO  Boolean for running protocol with AbInitioVO score function, runs with flag
-refinesubset  --Refine_Subset  Only subjects the lowest X% of structures to Relax refinement where X is specified as input following flag. Default 100
-relnstruct  --Number_of_Relax_Structures  Number of independent full-atom Relax sampling trajectories from a single AbInitio structure. Default 50
```
### Acquiring the Necessary Inputs
All of the files (FASTA, Disordered Probability Prediction and Fragment Libraries) can all be obtained from the protein sequence as detailed below
#### Disordered Probability Prediction
Disordered probability predictions can be obtained using [RaptorX Property](http://raptorx.uchicago.edu/StructurePropertyPred/predict/), [PsiPred DISOPRED3](http://bioinf.cs.ucl.ac.uk/psipred/) or other servers of your choice.
#### Fragment Library Construction
##### Secondary Structure Prediction
Secondary structure predictions can be obtained using [RaptorX Property](http://raptorx.uchicago.edu/StructurePropertyPred/predict/), [PsiPred PSIPRED 4.0](http://bioinf.cs.ucl.ac.uk/psipred/) or other servers of your choice.
After obtaining these prediction files, the script _ _Diso_SS2_Reweight_Opt_FFT.py_ _ can be used to generate secondary structure predictions files that are reweighted based on disordered probabilities and vice-versa.
- To generate a reweighted secondary structure prediction based on a disordered probability prediction:
	```run Diso_SS2_Reweight_Opt_FFT.py -ssin INPUT_SS2_PREDICTION_FILE -sstype 4_LETTER_CODE -disoin INPUT_DISORDER_PREDICTION_FILE -out REWEIGHTED_SS2_PREDICTION_OUTPUT_FILE```
- To generate a reweighted disordered probability based on the secondary structure prediction, add the following flag to the above run command:
	```-disoout REWEIGHTED_DISORDER_PREDICTION_OUTPUT_FILE```
Currently this code only supports inputs from disodered probability predictions from the servers named above (which take on the same format) and secondary structure predictions from those named above and [JUFO](http://www.meilerlab.org/index.php/servers/show?s_id=5),
which all vary in format. All secondary structure outputs follow the PSIPRED format rquired by the Rosetta [FragmentPicker](https://www.rosettacommons.org/docs/latest/application_documentation/utilities/app-fragment-picker) application.
## Running FastFloppyTail

## Analysis
